woocommerce/woocommerce-ios
# Fix: update text for Variation Image Placeholder

Closes: https://github.com/woocommerce/woocommerce-ios/issues/4848


### Description
When looking at a variation (within a variable product) with no images, the image placeholder states “Add a product image”. This PR allows us to detect when a variation is being viewed, and have the image placeholder state “Add a variation image” instead. 

This element was initially static, so this required connecting more data to the UI.
To do so, we defined a new boolean `isVariation` as `isVariation: product is EditableProductVariationModel`.
Then we passed the new boolean through various layers, and added a new configuration function in the cell, to define the label text when `isVariation` is true.

Finally, we added three unit tests:
1. In `ExtendedAddProductImageCollectionViewCellTests`, the new unit test checks whether the label text of a cell with `isVariation: true` is different from the label text of a cell with `isVariation: false`.
2. In `DefaultProductFormTableViewModelTests`, the new unit test checks whether, given a product variation, after the viewModel is built, the `.image` row has `isVariation: true`.
3. Also in `DefaultProductFormTableViewModelTests`, another new unit test checks whether, given a product (not a variation), after the viewModel is built, the `.image` row has `isVariation: false`.

Hat-tip @ealeksandrov for helping me structure this approach and answering all my questions along the way.  @rachelmcr for reminding me to check my assumptions about variations, resulting in a final breakthrough in the unit tests.


### Testing instructions
In the Products section, navigate to any product that doesn't have images yet.
Note the image label states “Add a product image”
Now navigate to a variation (not a _variable product_ - a _variation_ within a variable product) that doesn't have images yet.
Note the image label states “Add a variation image”

### Testing the unit tests
I've tested the unit tests in the following ways:
- Check that each test passes

For unit test 2 & 3:
- Change `0` to `1` in `guard case let .primaryFields(rows) = tableViewModel.sections[0]`. This will result in the error "Unexpected section at index 0".
- Pass a Product instead of a Variation, or vice-versa. I did this by swapping the `//Arrange` sections between the two tests. This will result in `XCTAssertTrue failed` (isVariation was false) and `XCTAssertFalse failed` (isVariation was true), respectively.
- Make sure we don't get the variable data.  One way to do this is to comment out `isVariation = isVariationValue` in `if case .images(_, _, let isVariationValue) = row {`, but it could also be done by breaking the view model so that `.image` isn't shown.
This will result in the error "Cell not found".

Can you think of any other situations in which the unit tests should fail? Please test them if so!

---
